toonz/sources/toonz/commandbarpopup.cpp

```diff
@@ -0,0 +1,480 @@
+            RightClickMenuCommandType);
+  addFolder(ShortcutTree::tr("Tool Modifiers"), ToolModifierCommandType);
+  addFolder(ShortcutTree::tr("Visualization"), ZoomCommandType);
+  addFolder(ShortcutTree::tr("Misc"), MiscCommandType);
+  addFolder(ShortcutTree::tr("RGBA Channels"), RGBACommandType);
+
+  CommandBarSeparatorItem* sep = new CommandBarSeparatorItem(0);
+  sep->setToolTip(0, QObject::tr("[Drag&Drop] to copy separator to menu bar"));
+  addTopLevelItem(sep);
+}
+
+//-----------------------------------------------------------------------------
+
+void CommandBarListTree::addFolder(const QString& title, int commandType,
+                                   QTreeWidgetItem* parentFolder) {
+  QTreeWidgetItem* folder;
+  if (!parentFolder)
+    folder = new QTreeWidgetItem(this);
+  else
+    folder = new QTreeWidgetItem(parentFolder);
+  assert(folder);
+  folder->setText(0, title);
+  folder->setIcon(0, invisibleRootItem()->icon(0));
+
+  std::vector<QAction*> actions;
+  CommandManager::instance()->getActions((CommandType)commandType, actions);
+  for (int i = 0; i < (int)actions.size(); i++) {
+    CommandBarCommandItem* item = new CommandBarCommandItem(folder, actions[i]);
+    item->setToolTip(0, QObject::tr("[Drag&Drop] to copy command to menu bar"));
+  }
+}
+
+//-----------------------------------------------------------------------------
+
+void CommandBarListTree::mousePressEvent(QMouseEvent* event) {
+  setCurrentItem(itemAt(event->pos()));
+  CommandBarCommandItem* commandItem =
+      dynamic_cast<CommandBarCommandItem*>(itemAt(event->pos()));
+  CommandBarSeparatorItem* separatorItem =
+      dynamic_cast<CommandBarSeparatorItem*>(itemAt(event->pos()));
+
+  if (commandItem || separatorItem) {
+    std::string dragStr;
+    QString dragPixmapTxt;
+    if (commandItem) {
+      dragStr =
+          CommandManager::instance()->getIdFromAction(commandItem->getAction());
+      dragPixmapTxt = commandItem->getAction()->text();
+      dragPixmapTxt.remove("&");
+    } else {
+      dragStr       = "separator";
+      dragPixmapTxt = tr("----Separator----");
+    }
+
+    QMimeData* mimeData = new QMimeData;
+    mimeData->setText(QString::fromStdString(dragStr));
+
+    QFontMetrics fm(QApplication::font());
+    QPixmap pix(fm.boundingRect(dragPixmapTxt).adjusted(-2, -2, 2, 2).size());
+    QPainter painter(&pix);
+    painter.fillRect(pix.rect(), Qt::white);
+    painter.setPen(Qt::black);
+    painter.drawText(pix.rect(), Qt::AlignCenter, dragPixmapTxt);
+
+    QDrag* drag = new QDrag(this);
+    drag->setMimeData(mimeData);
+    drag->setPixmap(pix);
+
+    drag->exec(Qt::CopyAction);
+  }
+
+  QTreeWidget::mousePressEvent(event);
+}
+
+//=============================================================================
+// CommandBarPopup
+//-----------------------------------------------------------------------------
+
+CommandBarPopup::CommandBarPopup(bool isXsheetToolbar)
```

Comment:
It seems that `CommandBarPopup` overlaps in many part with `MenuBarPopup` .
Is there a possibility to share the same code, in order to save the maintenance cost?

Reply:
I took a look at the files and while the basics are very similar (I used MenuBarPopup to make CommandBarPopup), most of the functions have small changes that make sharing the same code difficult.  If they both inherited from a common base class, most functions would need to be overridden anyway.  I'm sure there is a way to do it, but honestly with this school year being as busy as it is, I wouldn't have time to take it on right now.

Reply:
No problem, then let's just leave it for the future issue.